applications/crossbar/src/api_util.erl

```diff
@@ -563,8 +554,30 @@ is_valid_request_envelope(Envelope, Context) ->
 
 -spec requires_envelope(cb_context:context()) -> boolean().
 requires_envelope(Context) ->
+    Routines = [fun api_version_requires_envelope/1
+               ,fun content_type_requires_envelope/1
+               ,fun req_noun_requires_envelope/1
+               ],
+    lists:all(fun(F) -> F(Context) end, Routines).
+
+-spec api_version_requires_envelope(cb_context:context()) -> boolean().
+api_version_requires_envelope(Context) ->
     not lists:member(cb_context:api_version(Context), ?NO_ENVELOPE_VERSIONS).
 
+-spec content_type_requires_envelope(cb_context:context()) -> boolean().
+content_type_requires_envelope(Context) ->
+    not lists:member(cb_context:req_header(Context, <<"content-type">>), ?NO_ENVELOPE_CONTENT_TYPES).
+
+-spec req_noun_requires_envelope(cb_context:context()) -> boolean().
+req_noun_requires_envelope(Context) ->
+    req_noun_requires_envelope(Context, cb_context:req_nouns(Context)).
+
+-spec req_noun_requires_envelope(cb_context:context(), req_nouns()) -> boolean().
+req_noun_requires_envelope(_Context, []) -> 'true';
+req_noun_requires_envelope(Context, [{Module, Params} | _]) ->
+    kz_module:is_exported(Module, 'requires_envelope', length(Params)+1)
+        andalso erlang:apply(Module, 'requires_envelope', [Context | Params]).
```

Comment:
Doesn't `Module` come in as a binary? Wouldn't this require a `kz_term:to_atom/1` call?

Reply:
fixed